pulsar/log/logger.go

```diff
@@ -0,0 +1,57 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package log
+
+// Logger and Entry interfaces are inspired by sirupsen/logrus
+
+// Fields type, used to pass to `WithFields`.
+type Fields map[string]interface{}
+
+// Logger describes the interface that must be implemeted by all loggers.
+type Logger interface {
+	SubLogger(fields Fields) Logger
+
+	WithFields(fields Fields) Entry
+	WithField(name string, value interface{}) Entry
+
+	Debug(args ...interface{})
+	Info(args ...interface{})
+	Warn(args ...interface{})
+	Error(args ...interface{})
+
+	Debugf(format string, args ...interface{})
+	Infof(format string, args ...interface{})
+	Warnf(format string, args ...interface{})
+	Errorf(format string, args ...interface{})
+}
+
+// Entry describes the interface for the logger entry.
+type Entry interface {
```

Comment:
Is it possible to leave one interface? E.g. `SubLogger(fields Fields) Logger` is semantically close to  `WithFields(fields Fields) Entry`, so probably we can leave just `Entry`?

Reply:
Yeah, `Logger` and `Entry` share many methods, but their responsibilities are different. In general, `Entry` just represents one log entry, once fired it should be released. Separating entry from logger enables users to fine-grained control the implementation, such as pooling the entries. I'm not sure whether it's more friendly and easier to use by combining the two interfaces into one.

Reply:
Currently there is no pooling, so having two entities just complicates implementation and slightly confuses.
The `SubLogger(fields Fields)` result is technically `logrus.Entry`.

Reply:
For `logrus` backed implementation, it's true, but the logger interfaces are not targeted only for logrus, they are more for users to customize the implementation to cope with their own scenarios. 

Using two entities makes logging concepts clear and gives users more control on the implementation, in some cases,  logging performance is critical and entry pooling is necessary. Logger is usually used to hold the log level, common fields and its own entry pool, where the entry holds its private fields, formats all its fields into bytes and emits the bytes out.  I think implementation based on these concepts is much straightforward. The popular `logrus` provides both logger-level and entry-level methods, but they are not based on interfaces. Another the performant logging library `zap` provides only Logger level methods, but all loggers share a global entry pool.

Reply:
Agree, but it seems like logrus approach with fields (`map[string]interface{}`) will reduce benefits from pooling.

Moreover, attempt to abstract logging is already limits performance optimizations due to usage of interfaces instead of structures, so I'm not sure that having both interfaces will simplify something performance-related.

There is another point: we are pre-v1, so it is OK to break backward compatibility at some point. If one decide to switch to zap or zerolog and employ it's performance benefits, they will be forced to change interfaces (but I still think that it is nearly impossible to create both abstracted and performant logger interfaces)